File: android/app/src/main/java/com/costoda/dittoedgestudio/data/logging/LogPatternStore.kt
package com.costoda.dittoedgestudio.data.logging

import com.costoda.dittoedgestudio.domain.model.LogPattern
import com.costoda.dittoedgestudio.domain.model.LogPatternBody
import com.costoda.dittoedgestudio.domain.model.PatternSource
import com.costoda.dittoedgestudio.domain.model.parseLevelFilter
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Loads, validates and persists the log-analysis pattern catalog (parity port of
 * the VS Code extension's `PatternStore`).
 *
 * Sources:
 * - **Bundled** — `assets/problem_patterns.json` (read-only catalog of known
 *   Ditto problem signatures).
 * - **User** — `<filesDir>/log-analyzer/user_patterns.json`, same JSON shape.
 *
 * A user pattern cannot reuse a bundled key. Invalid entries are dropped from
 * [patterns] and reported via [patternErrors]; a missing/corrupt user file is
 * tolerated (empty catalog).
 */
class LogPatternStore(
    private val userPatternsFile: File,
    private val bundledJsonLoader: () -> String?,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {

    // Declared before _patterns: the initial loadAll() publishes errors.
    private val _patternErrors = MutableStateFlow<Map<String, String>>(emptyMap())
    /** key → rejection reason for entries that failed validation. */
    val patternErrors: StateFlow<Map<String, String>> = _patternErrors.asStateFlow()

    private val _patterns = MutableStateFlow(loadAll())
    val patterns: StateFlow<Map<String, LogPattern>> = _patterns.asStateFlow()

    val bundledKeys: Set<String> get() = bundledBodies().keys

    private fun bundledBodies(): Map<String, LogPatternBody> =
        bundledJsonLoader()
            ?.let { raw -> runCatching { json.decodeFromString<Map<String, LogPatternBody>>(raw) }.getOrNull() }
            ?: emptyMap()

    private fun userBodies(): Map<String, LogPatternBody> {
        if (!userPatternsFile.exists()) return emptyMap()
        return runCatching {
            json.decodeFromString<Map<String, LogPatternBody>>(userPatternsFile.readText())
        }.getOrElse {
            emptyMap()
        }
    }

    private fun loadAll(): Map<String, LogPattern> {
        val errors = mutableMapOf<String, String>()
        val out = linkedMapOf<String, LogPattern>()

        fun addAll(bodies: Map<String, LogPatternBody>, source: PatternSource) {
            for ((key, body) in bodies) {
                val reason = LogPatternEngine.rejectReason(key, body, source)
                if (reason != null) {
                    errors[key] = reason
                    continue
                }
                out[key] = LogPattern(
                    key = key,
                    body = body,
                    severity = body.severity,
                    levelFilter = parseLevelFilter(body.levelFilter),
                    source = source,
                )
            }
        }

        addAll(bundledBodies(), PatternSource.BUNDLED)
        // User patterns may override bundled by key only via hand-editing the file,
        // matching the extension; add()/update() forbid bundled-key collisions.
        addAll(userBodies(), PatternSource.USER)

        _patternErrors.value = errors
        return out
    }

    /** Re-reads both sources — call after external edits of the user file. */
    fun reload() {
        _patterns.value = loadAll()
    }

    suspend fun add(key: String, body: LogPatternBody) = withContext(Dispatchers.IO) {
        require(!bundledKeys.contains(key)) { "key collides with a bundled pattern" }
        require(!userBodies().containsKey(key)) { "a pattern with this key already exists" }
        writeUser(userBodies() + (key to body))
        reload()
    }

    suspend fun update(key: String, body: LogPatternBody) = withContext(Dispatchers.IO) {
        require(userBodies().containsKey(key)) { "pattern '$key' is not a user pattern" }
        writeUser(userBodies().toMutableMap().apply { put(key, body) })
        reload()
    }

    suspend fun delete(key: String) = withContext(Dispatchers.IO) {
        require(userBodies().containsKey(key)) { "pattern '$key' is not a user pattern" }
        writeUser(userBodies().toMutableMap().apply { remove(key) })
        reload()
    }

    private fun writeUser(bodies: Map<String, LogPatternBody>) {
        userPatternsFile.parentFile?.mkdirs()
        // Write-temp-then-rename: a mid-write crash must not leave a corrupt
        // catalog file behind (Swift port writes atomically; keep parity).
        val payload = json.encodeToString(bodies)
        val tmp = File(userPatternsFile.parentFile, "${userPatternsFile.name}.tmp")
        tmp.writeText(payload)
        if (!tmp.renameTo(userPatternsFile)) {
            tmp.copyTo(userPatternsFile, overwrite = true)
            tmp.delete()
        }
    }
}

File: SwiftUI/EdgeStudio/Data/LogPatternStore.swift
import DittoSwift
import Foundation

/// Loads, validates and persists the log-analysis pattern catalog (parity port
/// of the VS Code extension's `PatternStore`).
///
/// Sources:
/// - **Bundled** — `problem_patterns.json` in the app bundle (read-only catalog
///   of known Ditto problem signatures).
/// - **User** — `~/Library/Application Support/ditto_edge_studio/log-analyzer/user_patterns.json`.
///
/// A user pattern cannot reuse a bundled key. Invalid entries are dropped from
/// `patterns` and reported via `patternErrors`; a missing/corrupt user file is
/// tolerated (empty catalog).
@MainActor @Observable
final class LogPatternStore {
    /// Validated patterns by key (bundled + user).
    private(set) var patterns: [String: LogPattern] = [:]

    /// key → rejection reason for entries that failed validation.
    private(set) var patternErrors: [String: String] = [:]

    /// Bumped on every catalog change so views can re-run scans.
    private(set) var revision = 0

    var bundledKeys: Set<String> {
        Set(bundledBodies().keys)
    }

    private let bundledResourceURL: URL?
    private let userPatternsFileURL: URL

    /// - Parameters:
    ///   - bundledResourceURL: defaults to the bundled `problem_patterns.json`.
    ///   - userPatternsFileURL: defaults to the app-support `log-analyzer` location;
    ///     injectable for tests.
    init(
        bundledResourceURL: URL? = Bundle.main.url(forResource: "problem_patterns", withExtension: "json"),
        userPatternsFileURL: URL? = nil
    ) {
        self.bundledResourceURL = bundledResourceURL
        if let userPatternsFileURL {
            self.userPatternsFileURL = userPatternsFileURL
        } else {
            let appSupport = FileManager.default.urls(
                for: .applicationSupportDirectory,
                in: .userDomainMask
            )[0]
            self.userPatternsFileURL = appSupport
                .appendingPathComponent("ditto_edge_studio", isDirectory: true)
                .appendingPathComponent("log-analyzer", isDirectory: true)
                .appendingPathComponent("user_patterns.json")
        }
        reload()
    }

    enum StoreError: Error, LocalizedError {
        case collidesWithBundledKey(String)
        case patternAlreadyExists(String)
        case notAUserPattern(String)

        var errorDescription: String? {
            switch self {
            case let .collidesWithBundledKey(key):
                return "'\(key)' collides with a bundled pattern"
            case let .patternAlreadyExists(key):
                return "a pattern with key '\(key)' already exists"
            case let .notAUserPattern(key):
                return "'\(key)' is not a user pattern"
            }
        }
    }

    // MARK: - Loading

    private func decode(_ url: URL) -> [String: LogPatternBody] {
        guard let data = try? Data(contentsOf: url) else { return [:] }
        return (try? JSONDecoder().decode([String: LogPatternBody].self, from: data)) ?? [:]
    }

    private func bundledBodies() -> [String: LogPatternBody] {
        guard let bundledResourceURL else { return [:] }
        return decode(bundledResourceURL)
    }

    private func userBodies() -> [String: LogPatternBody] {
        decode(userPatternsFileURL)
    }

    /// Re-reads both sources — call after external edits of the user file.
    func reload() {
        var errors: [String: String] = [:]
        var out: [String: LogPattern] = [:]

        func addAll(_ bodies: [String: LogPatternBody], source: PatternSource) {
            for (key, body) in bodies {
                if let reason = LogPatternEngine.rejectReason(key: key, body: body, source: source) {
                    errors[key] = reason
                    continue
                }
                out[key] = LogPattern(
                    key: key,
                    body: body,
                    levelFilter: parseLogLevelFilter(body.levelFilter),
                    source: source
                )
            }
        }

        addAll(bundledBodies(), source: .bundled)
        // User patterns may override bundled by key only via hand-editing the file
        // (parity with the extension); add/update forbid bundled-key collisions.
        addAll(userBodies(), source: .user)

        patterns = out
        patternErrors = errors
        revision += 1
    }

    // MARK: - CRUD

    func add(key: String, body: LogPatternBody) throws {
        let resolvedKey = key.trimmingCharacters(in: .whitespaces)
        guard !bundledKeys.contains(resolvedKey) else {
            throw StoreError.collidesWithBundledKey(resolvedKey)
        }
        var users = userBodies()
        guard users[resolvedKey] == nil else {
            throw StoreError.patternAlreadyExists(resolvedKey)
        }
        users[resolvedKey] = body
        try writeUser(users)
        reload()
    }

    func update(key: String, body: LogPatternBody) throws {
        var users = userBodies()
        guard users[key] != nil else { throw StoreError.notAUserPattern(key) }
        users[key] = body
        try writeUser(users)
        reload()
    }

    func delete(key: String) throws {
        var users = userBodies()
        guard users.removeValue(forKey: key) != nil else {
            throw StoreError.notAUserPattern(key)
        }
        try writeUser(users)
        reload()
    }

    private func writeUser(_ bodies: [String: LogPatternBody]) throws {
        let directory = userPatternsFileURL.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let data = try JSONEncoder.prettySorted.encode(bodies)
        try data.write(to: userPatternsFileURL, options: .atomic)
    }
}

private extension JSONEncoder {
    static var prettySorted: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return encoder
    }
}
